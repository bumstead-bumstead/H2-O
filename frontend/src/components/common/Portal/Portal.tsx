import type { PropsWithChildren } from 'react';
import { useRef } from 'react';
import { createPortal } from 'react-dom';
import { useMounted } from '@/hooks';

interface Props {
  targetElementId?: string;
}

function Portal({ children, targetElementId = 'portal-root' }: PropsWithChildren<Props>) {
  const elementRef = useRef<HTMLElement | null>(null);
  const mounted = useMounted();

  elementRef.current = document.getElementById(targetElementId);

  if (!mounted || !elementRef.current) return null;

  return createPortal(children, elementRef.current);
}

export default Portal;
