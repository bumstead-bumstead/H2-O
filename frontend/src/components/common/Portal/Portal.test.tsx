import Portal from './Portal';
import { render, screen } from '@/tests/test-util';

describe('Portal 컴포넌트 테스트', () => {
  beforeEach(() => {
    const portalRoot = global.document.createElement('div');
    portalRoot.setAttribute('id', 'portal-root');

    const body = global.document.querySelector('body');
    if (!body) throw new Error('portal-root의 body를 찾을 수 없습니다.');
    body.appendChild(portalRoot);
  });

  it('Portal이 정상적으로 렌더링 된다.', () => {
    render(<Portal />);

    expect(document.getElementById('portal-root')).toBeInTheDocument();
  });

  it('Portal의 child가 정상적으로 렌더링 된다.', () => {
    render(
      <Portal>
        <span>Hello World</span>
      </Portal>,
    );

    expect(screen.getByText('Hello World')).toBeInTheDocument();
  });

  it('Portal 컴포넌트 스냅샷 테스트', () => {
    render(<Portal />);

    expect(document.getElementById('portal-root')).toMatchSnapshot();
  });
});
