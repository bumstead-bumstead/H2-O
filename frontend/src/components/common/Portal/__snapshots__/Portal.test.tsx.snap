// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Portal 컴포넌트 테스트 > Portal 컴포넌트 스냅샷 테스트 1`] = `
<div
  id="portal-root"
/>
`;
